src/fmtargs.h

```diff
@@ -0,0 +1,83 @@
+#define __NARG__(...)  __NARG_I_(__VA_ARGS__,__RSEQ_N())
+#define __NARG_I_(...) __ARG_N(__VA_ARGS__)
+
+/* Define a macro which will call an arbitrary macro appended with a number indicating
+ * the number of arguments it has. */
+#define _VFUNC_(name, n) name##n
+#define _VFUNC(name, n) _VFUNC_(name, n)
+#define VFUNC(func, ...) _VFUNC(func, __NARG__(__VA_ARGS__)) (__VA_ARGS__)
+
+#define COMPACT_FMT(...) VFUNC(COMPACT_FMT_, __VA_ARGS__)
+#define COMPACT_VALUES(...) VFUNC(COMPACT_VALUES_, __VA_ARGS__)
+#define FMT_STRING(initial_string, ...) initial_string COMPACT_FMT(__VA_ARGS__), COMPACT_VALUES(__VA_ARGS__)
+
+/* Everything below this line is automatically generated by utils/generate_fmtargs.py, do not manually edit */
```

Comment:
Should probably just auto-generate the entire file.

Reply:
i don't think so, since we're not aiming to really use this for **automatic** code generation, but rather a manual one (once in a blue moon, or never), i don't think we want to copy a big block of manually written code into that generator.

instead, i suggest adding more comments in the non-auto generated ones to make it easier to read.
starting with some header at the top that describes what it's good for, how to use it, and how it works.
e.g. `using that much pre-processor tricks to re-order argument for the plan printf function call, so that in printf("a=%d b=%s, arg1, arg2), we can put each argument next to it's format string portion.` or alike.